devel/fortio/periodic.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fortio
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Function to run periodically.
+type Function func(tid int)
+
+// RunnerOptions are the parameters to the PeriodicRunner.
+type RunnerOptions struct {
+	Function Function
+	QPS      float64
+	Duration time.Duration
+	// Note that this actually maps to gorountines and not actual threads
+	// but threads seems like a more familiar name to use for non go users
+	// and in a benchmarking context
+	NumThreads  int
+	Percentiles []float64
+	Resolution  float64
+	// Debug/Verbosity level (0 is quietest)
+	Verbosity int
+}
+
+// PeriodicRunner let's you exercise the Function at the given QPS and collect
+// statistics and histogram about the run.
+type PeriodicRunner interface {
+	// Starts the run
+	Run()
+	// Returns the options normalized by constructor - do not mutate
+	// (where is const when you need it...)
+	GetOptions() *RunnerOptions
+}
+
+// Unexposed implementation details for PeriodicRunner.
+type periodicRunner struct {
+	RunnerOptions
+}
+
+// internal version, returning the concrete implementation.
+func newPeriodicRunner(opts *RunnerOptions) *periodicRunner {
+	r := &periodicRunner{*opts} // by default just copy the input params
+	if r.QPS < 0 {
+		log.Printf("Negative qps %f means max speed mode/no wait between calls", r.QPS)
+		r.QPS = 0
+	}
+	if r.NumThreads == 0 {
+		r.NumThreads = 4 // default
+	}
+	if r.NumThreads < 1 {
+		r.NumThreads = 1
+	}
+	if r.Percentiles == nil {
+		r.Percentiles = []float64{90.0}
+	}
+	if r.Resolution <= 0 {
+		r.Resolution = 0.001 // millisecond default
+	}
+	return r
```

Comment:
Do you want to explicitly set `r.Duration` to a default here?  Also I recommend defining default values as constants at the top of the file